restricted_site_access.php

```diff
@@ -122,6 +130,16 @@ public static function add_actions() {
 
 		add_filter( 'pre_option_blog_public', array( __CLASS__, 'pre_option_blog_public' ), 10, 1 );
 		add_filter( 'pre_site_option_blog_public', array( __CLASS__, 'pre_option_blog_public' ), 10, 1 );
+
+		// Prevent WordPress from auto-resolving 404 URLs.
+		add_filter( 'do_redirect_guess_404_permalink', '__return_true' );
+	}
+
+	/**
+	 * Generates a nonce on init.
+	 */
+	public static function generate_nonce() {
+		self::$redirection_nonce = wp_create_nonce( 'redirection_nonce' );
 	}
```

Comment:
One thing to be aware of here is that the cookie only considers the first nonce in the current code (0 - 12 hour lifespan).

If someone is exceptionally unlucky, they may end up having the cookie set with a nonce that expires by the time their browser redirects. This would then cause the redirect to happen again.

Two redirects is better than 1/0 redirects so I think it's fine for the initial fix. If you disagree, I have suggestions :)

Reply:
I agree and I look forward to your suggestions.